loopy/library/function.py

```diff
@@ -38,12 +39,12 @@
 
 class MakeTupleCallable(ScalarCallable):
     def with_types(self, arg_id_to_dtype, callables_table):
-        new_arg_id_to_dtype = arg_id_to_dtype.copy()
+        new_arg_id_to_dtype = dict(arg_id_to_dtype)
         for i in range(len(arg_id_to_dtype)):
             if i in arg_id_to_dtype and arg_id_to_dtype[i] is not None:
                 new_arg_id_to_dtype[-i-1] = new_arg_id_to_dtype[i]
 
-        return (self.copy(arg_id_to_dtype=new_arg_id_to_dtype,
+        return (self.copy(arg_id_to_dtype=constantdict(new_arg_id_to_dtype),
```

Comment:
This could probably be smth like:

```python
        new_arg_id_to_dtype = constantdict(arg_id_to_dtype).mutate()
        for i in range(len(arg_id_to_dtype)):
            if i in arg_id_to_dtype and arg_id_to_dtype[i] is not None:
                new_arg_id_to_dtype[-i-1] = new_arg_id_to_dtype[i]

        return (self.copy(arg_id_to_dtype=new_arg_id_to_dtype.finish()),
```

This avoids a second copy.

Reply:
Ah, I see! Thanks a lot for the example. I'll make those changes.

Reply:
I think I got all the places where this made sense (at least in the parts that were already modified). Let me know if you spot anything else!